arch/x86_64/mcount-dynamic.c

```diff
@@ -141,17 +143,11 @@ static void read_xray_map(struct arch_dynamic_info *adi,
 	elf_get_secdata(elf, iter);
 	elf_read_secdata(elf, iter, 0, adi->xrmap, shdr->sh_size);
 
-	/* handle position independent code */
-	if (elf->ehdr.e_type == ET_DYN) {
-		struct xray_instr_map *xrmap;
-		unsigned i;
-
-		for (i = 0; i < adi->xrmap_count; i++) {
-			xrmap = &adi->xrmap[i];
+	for (i = 0; i < adi->xrmap_count; i++) {
+		struct xray_instr_map *xrmap = &adi->xrmap[i];
 
-			xrmap->addr  += offset;
-			xrmap->entry += offset;
-		}
+		xrmap->instr_addr += offset + (shdr->sh_offset + i * sizeof(struct xray_instr_map));
+		xrmap->func_begin_addr += offset + (shdr->sh_offset + i * sizeof(struct xray_instr_map) + 8);
```

Comment:
It seems this logic should be applied only when the X-ray sled entry version is 2 or above.